File: scripts/check_imports.js
#!/usr/bin/env node
/**
 * Скрипт для проверки импортов в frontend компонентах.
 * 
 * Проверяет:
 * - Использование api без импорта
 * - Использование useLocalStorage без импорта
 * - Использование useModels без импорта
 * - Использование useApi без импорта
 * 
 * Запуск: node scripts/check_imports.js
 */

const fs = require('fs');
const path = require('path');

const FRONTEND_DIR = path.join(__dirname, '..', 'frontend', 'src');
const COMPONENTS_DIR = path.join(FRONTEND_DIR, 'components');
const HOOKS_DIR = path.join(FRONTEND_DIR, 'hooks');
const UTILS_DIR = path.join(FRONTEND_DIR, 'utils');

const ISSUES = [];

// Паттерны для поиска использования без импорта
const PATTERNS = [
  { name: 'api', pattern: /\bapi\.(models|conversations|projects|metrics|code|settings|tasks|databases|stream)\s*\(/g, importPattern: /import\s+.*\bapi\b.*from\s+['"]\.\.\/services\/apiClient['"]/ },
  { name: 'useLocalStorage', pattern: /\buseLocalStorage\s*\(/g, importPattern: /import\s+.*\buseLocalStorage\b.*from\s+['"]\.\.\/hooks\/useLocalStorage['"]/ },
  { name: 'useLocalStorageString', pattern: /\buseLocalStorageString\s*\(/g, importPattern: /import\s+.*\buseLocalStorageString\b.*from\s+['"]\.\.\/hooks\/useLocalStorage['"]/ },
  { name: 'useModels', pattern: /\buseModels\s*\(/g, importPattern: /import\s+.*\buseModels\b.*from\s+['"]\.\.\/hooks\/useModels['"]/ },
  { name: 'useApi', pattern: /\buseApi\s*\(/g, importPattern: /import\s+.*\buseApi\b.*from\s+['"]\.\.\/hooks\/useApi['"]/ },
];

function checkFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf-8');
  const relativePath = path.relative(FRONTEND_DIR, filePath);
  
  // Пропускаем файлы, которые определяют эти функции
  if (filePath.includes('apiClient.ts') || 
      filePath.includes('useLocalStorage.ts') || 
      filePath.includes('useModels.ts') || 
      filePath.includes('useApi.ts')) {
    return;
  }
  
  PATTERNS.forEach(({ name, pattern, importPattern }) => {
    const matches = content.match(pattern);
    if (matches && matches.length > 0) {
      // Проверяем наличие импорта
      if (!importPattern.test(content)) {
        ISSUES.push({
          file: relativePath,
          issue: `Используется ${name} без импорта`,
          line: content.split('\n').findIndex(line => pattern.test(line)) + 1
        });
      }
    }
  });
}

function walkDir(dir, callback) {
  const files = fs.readdirSync(dir);
  
  files.forEach(file => {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    
    if (stat.isDirectory()) {
      walkDir(filePath, callback);
    } else if (file.endsWith('.ts') || file.endsWith('.tsx')) {
      callback(filePath);
    }
  });
}

// Проверяем компоненты
walkDir(COMPONENTS_DIR, checkFile);

// Проверяем хуки
walkDir(HOOKS_DIR, checkFile);

// Проверяем утилиты
walkDir(UTILS_DIR, checkFile);

// Проверяем App.tsx
const appPath = path.join(FRONTEND_DIR, 'App.tsx');
if (fs.existsSync(appPath)) {
  checkFile(appPath);
}

// Выводим результаты
if (ISSUES.length === 0) {
  console.log('✅ Все импорты корректны!');
  process.exit(0);
} else {
  console.error('❌ Найдены проблемы с импортами:\n');
  ISSUES.forEach(({ file, issue, line }) => {
    console.error(`  ${file}:${line} - ${issue}`);
  });
  process.exit(1);
}

File: scripts/check_imports.py
#!/usr/bin/env python3
"""
Скрипт для проверки импортов в frontend компонентах.

Проверяет:
- Использование api без импорта
- Использование useLocalStorage без импорта
- Использование useModels без импорта
- Использование useApi без импорта

Запуск: python scripts/check_imports.py
"""
import os
import re
from pathlib import Path

FRONTEND_DIR = Path(__file__).parent.parent / "frontend" / "src"
COMPONENTS_DIR = FRONTEND_DIR / "components"
HOOKS_DIR = FRONTEND_DIR / "hooks"
UTILS_DIR = FRONTEND_DIR / "utils"

ISSUES = []

# Паттерны для поиска использования без импорта
PATTERNS = [
    {
        "name": "api",
        "pattern": r"\bapi\.(models|conversations|projects|metrics|code|settings|tasks|databases|stream)\s*\(",
        "import_pattern": r"import\s+.*\bapi\b.*from\s+['\"][\.\/]*services\/apiClient['\"]",
    },
    {
        "name": "useLocalStorage",
        "pattern": r"\buseLocalStorage\s*\(",
        "import_pattern": r"import\s+.*\buseLocalStorage\b.*from\s+['\"][\.\/]*hooks\/useLocalStorage['\"]",
    },
    {
        "name": "useLocalStorageString",
        "pattern": r"\buseLocalStorageString\s*\(",
        "import_pattern": r"import\s+.*\buseLocalStorageString\b.*from\s+['\"][\.\/]*hooks\/useLocalStorage['\"]",
    },
    {
        "name": "useModels",
        "pattern": r"\buseModels\s*\(",
        "import_pattern": r"import\s+.*\buseModels\b.*from\s+['\"][\.\/]*hooks\/useModels['\"]",
    },
    {
        "name": "useApi",
        "pattern": r"\buseApi\s*\(",
        "import_pattern": r"import\s+.*\buseApi\b.*from\s+['\"][\.\/]*hooks\/useApi['\"]",
    },
]


def check_file(file_path: Path) -> None:
    """Проверяет файл на наличие использования без импорта."""
    content = file_path.read_text(encoding="utf-8")
    relative_path = file_path.relative_to(FRONTEND_DIR)
    
    # Пропускаем файлы, которые определяют эти функции
    if any(
        name in str(file_path)
        for name in ["apiClient.ts", "useLocalStorage.ts", "useModels.ts", "useApi.ts"]
    ):
        return
    
    for pattern_info in PATTERNS:
        name = pattern_info["name"]
        pattern = re.compile(pattern_info["pattern"])
        import_pattern = re.compile(pattern_info["import_pattern"])
        
        if pattern.search(content):
            # Проверяем наличие импорта
            if not import_pattern.search(content):
                # Находим строку с использованием
                lines = content.split("\n")
                for i, line in enumerate(lines, 1):
                    if pattern.search(line):
                        ISSUES.append(
                            {
                                "file": str(relative_path),
                                "issue": f"Используется {name} без импорта",
                                "line": i,
                            }
                        )
                        break


def walk_dir(directory: Path, callback) -> None:
    """Рекурсивно обходит директорию и вызывает callback для каждого .ts/.tsx файла."""
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith((".ts", ".tsx")):
                file_path = Path(root) / file
                callback(file_path)


# Проверяем компоненты
if COMPONENTS_DIR.exists():
    walk_dir(COMPONENTS_DIR, check_file)

# Проверяем хуки
if HOOKS_DIR.exists():
    walk_dir(HOOKS_DIR, check_file)

# Проверяем утилиты
if UTILS_DIR.exists():
    walk_dir(UTILS_DIR, check_file)

# Проверяем App.tsx
app_path = FRONTEND_DIR / "App.tsx"
if app_path.exists():
    check_file(app_path)

# Выводим результаты
if not ISSUES:
    print("✅ Все импорты корректны!")
    exit(0)
else:
    print("❌ Найдены проблемы с импортами:\n")
    for issue in ISSUES:
        print(f"  {issue['file']}:{issue['line']} - {issue['issue']}")
    exit(1)
